File: 0006-zigzag-conversion/solution.py
class Solution(object):
    def convert(self, s, numRows):
        if numRows == 1:
            return s

        zigZag = ["" for __ in xrange(numRows)]
        index = 0
        col = 0
        while col < len(s):
            while index <= numRows - 1 and col < len(s):
                zigZag[index] += s[col]
                index += 1
                col += 1
            index = numRows - 2
            while index >= 0 and col < len(s):
                zigZag[index] += s[col]
                col += 1
                index -= 1
            index = 1
                
        result = ""
        for row in zigZag:
            result += row
    
        return result
            
                

File: 0006-zigzag-conversion/solution.cpp
class Solution {
public:
    string convert(string s, int numRows) 
    {
        vector<string> zigZag(numRows, "");
        const int n = s.size();
        int i = 0;
        while(i < n)
        {
            int numSteps = 0;
            // down
            while(i < n && numSteps < numRows)
            {
                zigZag[numSteps] += s[i];
                i++;
                numSteps++;
            }
            //up
            numSteps = numRows-2;
            while(i < n && numSteps >= 1)
            {
                zigZag[numSteps] += s[i];
                i++;
                numSteps--;
            }
        }
        
        string ans = "";
        for(int i = 0; i < numRows; i++)
        {
            ans += zigZag[i];
        }
        
        return ans;
    }
};
